### PR TITLE
chore: Port Python tests to unittest

### LINKED ISSUE #2399: Use unittest in python tests
Originally we didn't have many tests in python. Deno is mostly TypeScript and Rust, so that makes sense. Why do we have python at all?

Python is used during build time for miscellaneous tasks and to run the deno and rust tests.

We use Python because V8/GN/Chrome uses Python. (Specifically Python 2.7 - we do not have the possibility of using Python 3 until Chromium makes the switch.)

Over time the Python code base has grown. We should try to be more sanitary in our testing of this internal code by using a real test framework: https://docs.python.org/2/library/unittest.html

A few places where unittest should be applied:
- [ ] [tools/util_test.py](https://github.com/denoland/deno/blob/1386b6ead586da15d0285941f2d980acc1d75eee/tools/util_test.py)
- [ ] [tools/setup_test.py](https://github.com/denoland/deno/blob/1386b6ead586da15d0285941f2d980acc1d75eee/tools/setup_test.py)
- [ ] [tools/permission_prompt_test.py](https://github.com/denoland/deno/blob/1386b6ead586da15d0285941f2d980acc1d75eee/tools/permission_prompt_test.py)
- [ ] [tools/complex_permissions_test.py](https://github.com/denoland/deno/blob/1386b6ead586da15d0285941f2d980acc1d75eee/tools/complex_permissions_test.py)

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes #2399

### CODE REVIEW

File: tools/util_test.py [hunk 1]
```diff
@@ -1,76 +1,78 @@
...
-    parse_exit_code_test()
-    shell_quote_win_test()
-    parse_wrk_output_test()
+    suite = unittest.TestLoader().loadTestsFromTestCase(TestUtil)
+    result = unittest.TextTestRunner(verbosity=2).run(suite)
+    if not result.wasSuccessful():
+        sys.exit(1)
 
 
 if __name__ == '__main__':
```
Comment (ry) [inline tools/util_test.py]: In this file and the others do this:
```py
if __name__ == '__main__':
  unittest.main()
```

File: tools/util_test.py [hunk 2]
```diff
@@ -1,76 +1,78 @@
...
 
 def util_test():
-    pattern_match_test()
-    parse_exit_code_test()
-    shell_quote_win_test()
-    parse_wrk_output_test()
+    suite = unittest.TestLoader().loadTestsFromTestCase(TestUtil)
+    result = unittest.TextTestRunner(verbosity=2).run(suite)
+    if not result.wasSuccessful():
+        sys.exit(1)
```
Comment (ry) [inline tools/util_test.py]: I don't think this function is necessary anymore...

File: tools/test.py [hunk 3]
```diff
@@ -68,6 +70,7 @@ def main(argv):
     # Python/build tools testing
     setup_test()
     util_test()
```
Comment (ry) [inline tools/test.py]: Replace these calls with `unittest.main()`

**Comment by bartlomieju** [comment]
@ry I tried doing `unittest.main()` in `test.py` but... we'd have to manually check if tests were successful:

```python
// test.py
...
deno_exe = os.path.join(build_dir, "deno" + executable_suffix)
check_exists(deno_exe)

# Python/build tools testing
unittest.loadTestsFromNames("setup_test", "util_test")
test_program = unittest.main(exit=False)
if not test_program.result.wasSuccessful():
    sys.exit(1)
...
```

I assumed current solution is cleaner, let me know what you think.

> Please check by randomly adjusting some asserts here and there that they still cause tools/test.py to fail. I will too later.

Did it, with current solution on test failure `test.py` exits with code 1

**Comment by hayd** [comment]
I wonder if you could do away with test.py and rely on discover (or pytest etc) https://stackoverflow.com/a/15630454/1240268 (or have test.py call discover?) It feels old school to use loadTestsFromNames and wasSuccessful explicitly!

If you move the building of methods to `__init__` and server set up to setUp and tearDown this should remove a lot of boilerplate. .. I think you can avoid the loadTest stuff. The benefit of that is that these test files could then be run individually. You can also pass verbosity etc as a cli , instead of hardcoding it.

Of course, we could move to this incrementally and this is progress!

**Comment by bartlomieju** [comment]
@hayd, I've checked discovery, but that would still require to exclude `complex_permissions_test.py` and `permission_prompt_test.py` - they need a bit of custom setup to properly pass `deno_exe` to tests.

Also adding another command just to run 4 files with Python unit tests seems like an overkill.

Let me know what you think

**Comment by hayd** [comment]
> they need a bit of custom setup to properly pass deno_exe to tests.

The way I'd do this is with a mixin util to add in deno_exe property (looked up by --release flag or whatever the logic is) or just a function. I'm happy to play with this later, once this is merged.

**Comment by hayd** [comment]
FYI I am iterating a little on this.

**Pull request merged**